### ISSUE TITLE
Screencast: send an audio stream to capture alongside video

### ISSUE THREAD

**Issue by orowith2os** [flatpak/xdg-desktop-portal#957]
Currently, afaik, the screencast portal only supports video capture. This means that when an application wants to capture audio as well, like Discord or a browser, it needs to have a separate dialog (afaict) and permissions to access all of PipeWire.

Allowing the screencast portal to send some audio from the captured application can make the process of screen sharing easier, as users don't need to select audio capture separately. This means that an end user can capture a game with OBS, for example, and it'll automatically give OBS the appropriate audio sources.

**Comment by matthiasclasen** [comment]
We've discussed this a bit on irc, and had the following thoughts:

- Ideal UX would be a "Also share audio from this window" checkbox when starting a screen cast
- pipewire should really know what audio and video streams are associated
- That might require assistance from the app in multi-window scenarios
- The simple 1-window-1-audio-stream case should be easy to guess based on pid

**Comment by wtay** [comment]
There is nothing to help you group audio and video in pipewire except for guessing based on the application.process.id.

I don't know if you can add metadata to windows in wayland.. if you can, you can set a metadata identifier on both the window and the associated stream. It would require application changes and It's only going to give better results for complicated cases where there are multiple windows and streams and only some are related.

**Comment by orowith2os** [comment]
Maybe it would just be best to add an additional menu that lets the user configure the audio sources to be sent too, then? linking PipeWire and Wayland this way definitely sounds like it would be a pain.

**Comment by dimtpap** [comment]
I wrote about some ways I used for adding app audio capture functionality to OBS Studio with PipeWire [here](https://github.com/dimtpap/obs-pipewire-audio-capture/wiki/Using-PipeWire-to-target-and-list-applications). Of course they're not perfect since as mentioned above, there is no relation between audio and video streams.

**Comment by orowith2os** [comment]
@mikepavone this sounds *very* finicky, and prone to breakage - would Discord be fine with just trusting the portals, and taking audio input from them? That would probably be the only reasonable way to handle it, unless you want to do yourself inside of Discord.... Which would mean full access to PipeWire, which is a *huge* no-no. 

There's also the world-is-beautiful scenario where applications associated their audio streams with windows on their own, and we all have our work cut out for us. But that isn't going to happen.

**Comment by mikepavone** [comment]
@orowith2os I mentioned the bit about pid based matching not because it's what we plan to do when using xdg-desktop-portal (it seems to be impossible even in the non-sandboxed case since the screencast API does not give any info for figuring out what application was selected), but in case it was helpful for the backend implementation.

We're fine with letting the portal implementation pick the stream that corresponds with the chosen window. Ideally this would be a single stream that is a mix of all the application's streams, including dealing with the child process issue, but we can handle mixing ourselves if needed as long as we get access to all the relevant streams. If you only handle the simple case of applications with a single stream where the same process owns the target window and audio stream I think that will result in a pretty broken experience for many common uses of screenshare on Discord.

If you end up with an implementation that requires the user to separately pick what application to stream audio from we will probably not ship an implementation of soundshare for xdg-desktop-portal. It's unclear to me from your comment if that's something being considered or not, but in case it is I figure I'll be clear. We don't feel any need to be the ones that figures out which audio streams are relevant, but we don't want to ship an implementation with poor UX if you decide that there is not a robust-enough solution on the backend.

Ultimately what we care about is being able to ship an experience on Wayland that is similar in capability and ease of use to what we ship elsewhere. We can be flexible about the details of how that is achieved.

**Comment by jadahl** [comment]
I agree that manual selection is not ideal, and we should not do that, providing audio output on a per-app granularity seems a saner to me. I think the "selecting" mentioned was about selecting *within* an application, but I don't think that is necessarily a good idea either.

> Ideally this would be a single stream that is a mix of all the application's streams, including dealing with the child process issue, but we can handle mixing ourselves if needed as long as we get access to all the relevant streams.

Couple of thoughts about mixing before or after handing things to an app:
 - Mixing before gives us the ability to just "fix" things if we end up being able to match an exact window with an audio stream. It also means apps cant "fix" things.
 - Mixing after gives the app the ability to "fix" things by selectively muting the "wrong" audio streams. It also means apps forever need to handle the mixing, even if it'll eventually be "fixed" in the portal/captured apps.

What kind of granularity does this feature have on other platforms? How often does an audio stream match 1 to 1 with a window, with a multi window multi audo stream application?

**Comment by GeorgesStavracas** [comment]
I also don't think any solution that involves users manually selecting audio streams would be acceptable.

From an API design perspective, I think there is precisely one thing that apps might need: a way to request audio sharing for screencasting. More concretely, that means apps should only pass a new boolean `audio` field to the [`SelectSources()` method](https://github.com/flatpak/xdg-desktop-portal/blob/main/data/org.freedesktop.portal.ScreenCast.xml#L75-L173) if they are capable of, and willing to, handle audio streams. (The naming of this field is still TBD.)

Then the host system will figure out which PipeWire streams match a certain window / monitor / virtual screen, and these streams will be exposed in the PipeWire remote that is created by the [`OpenPipeWireRemote()` method](https://github.com/flatpak/xdg-desktop-portal/blob/main/data/org.freedesktop.portal.ScreenCast.xml#L276-L294).

How this matching will happen on the host system is still to be agreed on; the current direction seems to be about adding more metadata keys to PipeWire, and a new Wayland protocol. This is still up to debate, but this won't impact in any meaningful way how this API works.

I still need to collect my thoughts about exposing a single mixed stream for each app window. But my initial reaction is that it'll be easier to expose all streams of each app individually, and let apps do mixing if so they wish.

**Comment by mikepavone** [comment]
> Mixing after gives the app the ability to "fix" things by selectively muting the "wrong" audio streams. It also means apps forever need to handle the mixing, even if it'll eventually be "fixed" in the portal/captured apps.

It's fairly uncommon for audio streams to be usefully distinguishable in apps that have more than one in my experience. Discord itself is a notable exception here, but it's mostly an accident of how our desktop app is put together. That said, I guess I see how this could be handy for folks setting up more complex scenarios with OBS and the like

> What kind of granularity does this feature have on other platforms? How often does an audio stream match 1 to 1 with a window, with a multi window multi audo stream application?

In general we only have process (or process tree) granularity for soundshare. I don't have any data on how often this provides a perfect match with user intent, but in practice it's not something we've seen many complaints about. In general, Discord screenshare seems to mostly be used for games and browser windows. For the first of those, the owning process of the window almost always owns all the audio streams for that window and none for other windows. For the second, it's much messier. Generally a child process will own all the audio streams for all tabs across all windows. For the browser case, generally what people really want to do is share a particular tab for both video and audio, but it doesn't seem to have been a huge issue in practice that we only have window granularity for video and process granularity for audio.

> I still need to collect my thoughts about exposing a single mixed stream for each app window. But my initial reaction is that it'll be easier to expose all streams of each app individually, and let apps do mixing if so they wish.

This is not a big deal for us since we need to be able to handle mixing for X11+Pulse anyway, but I think in practice most users of this API will probably want a mix. I suppose it could be argued that mixing multiple streams should be a generic pipewire feature though rather than something baked into this API.

**Comment by orowith2os** [comment]
> I also don't think any solution that involves users manually selecting audio streams would be acceptable.

Aside from once a "proper" solution is worked out, having this as a fallback (or optional path for users to take) would be very much appreciated, in the event that the proposed solutions fail to work for all use cases. From what @mikepavone said regarding browsers, that sounds like an appropriate use case for it. That, or allowing apps to have full access to PipeWire, so they can mix and match as they see fit - you can probably see why I want that as part of the portal instead.

Keeping in mind that I'm speaking mostly from an end user perspective - the internal details seem to all be very messy, and I can't quite make sense of them yet. I'm trying to make do as we go - I'll give my thoughts whenever (they're wanted) in the meantime.

**Comment by aaomidi** [comment]
I don’t think having the user pick the audio stream separately is a major issue. I would actually say it’s a feature rather than a bug to be able to do that. 

This would allow creative sound mixing when screaming. It also fits the “feel” of Linux as well.

I don’t think we should make perfect be the enemy of good here - an initial implementation that exposes the choice to the user can, in the future, do the automatic stream selection if a video and audio stream mapping ever became available.

**Comment by aaomidi** [comment]
@Mikenux 100% agreed, it's not a good experience (especially multiple times a day)

However, a few counter points:

1. Linux users currently _don't have a solution at all_. So anything that enables this is going to be a huge delivery anyway. Basically what I mentioned in my earlier comment of not letting perfect be the enemy of good, or even the enemy of "meh".

2. The audio stream selection is not "publicly exposed" in the API to applications like discord in the downstream. Improvements here can happen without breaking applications that are relying on the desktop-portal APIs.

3. Linux users are somewhat used to "not super great" user experiences - I suspect that some users would actually find being able to manually select the audio stream to be a feature rather than a bug.

4. Potentially basic fuzzy string similarity matching on the process information can maybe move the "likely stream" to the top of the list (this is me assuming that the process name is available in some capacity).

**Comment by Mikenux** [comment]
Yes, indeed, especially your fourth point. If mixing is wanted, then a user interface for selecting an audio source without capturing the window will also be needed.

But it depends on the current progress of the development of the feature as well as the design of the user interface. Only those (if there is someone) who are in charge of developing this can tell you.

**Comment by rcombs** [comment]
I'll note from the Discord end of things that we've now shipped screenshare support on Wayland using the screencast portal, and this issue is indeed preventing us from implementing audio sharing there. We'd still strongly support an API that provides pre-mixed audio (either process-and-children for windowed share, or all-minus-process-and-children for full-screen), but could implement mixing internally if an API emerged that required that. Having the user select audio streams separately from video continues not to be an option we would pursue.

**Comment by orowith2os** [comment]
Not having the user select the audio stream doesn't seem like it would be
your problem, because all you'd see is an audio stream, or not, on your
end, as a client.

The desktop gets to decide how that's done, whether it's waiting for
audio-video association protocols, matching audio streams to processes, or
having the user manually mix the audio streams.

**Comment by rcombs** [comment]
Ultimately, from the user's perspective, the app is responsible for the end-to-end experience; if the OS API used presents a poor UX, users perceive that as the *app* presenting a poor UX (which results in negative user sentiment and increased support burden, particularly as Linux gaming becomes more popular on emerging platforms). If the interface exposed by Portals is ultimately unable to provide the straightforward audio experience that users expect (unobtrusively-optional advanced flows notwithstanding), we'll likely need to pursue other routes for soundshare support in the future, if the feature is to be implemented at all.

**Comment by Mikenux** [comment]
Currently, the thing is to have:
- A video source picker.
- An audio source picker.
- An app source picker with matching audio (per app window), with the option for the user to not have the audio.

Concretely, what is missing?

**Comment by Mikenux** [comment]
I would say that default app audio = any audio that is not linked to any audio input source (device or app). This way, when an app has sources connected to input sources, we can present which audio sources to select.

**Comment by YoJames2019** [comment]
sooooooooooo is this just dead?

**Comment by polyv0x** [comment]
Great suggestion IMO! 

I've been wrestling with this myself in Linux and found that there currently also isn't a very reliable way to match say a Pulse Audio sink in to a chosen window. (There isn't enough overlapping data to match perfectly it seems, imperfect approximation using heuristics is possible though.) Not to mention the pains of dealing with an unsynced audio stream with a separate video stream. It would require dealing with 2 independent clocks. Playing them side by side would certainly result in desync. Or at the very least require more juggling to get it properly synced up.

> Hi [@GeorgesStavracas](https://github.com/GeorgesStavracas) ! I would like to work on this issue. Do you still think exposing individual streams is the better approach, or should we aim to provide a single mixed stream?

So my vote definitely goes to pre-mixed streams. (especially given the context that games have multiple sometimes, you wouldn't want to capture only a single audio stream from a game and miss the rest for example. And you wouldn't _in most cases_ want to put that burden on the user I would say.)

MacOS has this behavior standardized already, where selecting a single window with the native picker, also includes that window's audio. I think that would be a good example from how it's supposed to work from a user perspective.

You select a window, and whatever it is that it gets sent to, can choose to use the bundled audio stream or not.

**Comment by Nuhiat-Arefin** [comment]
> Great suggestion IMO!
> 
> I've been wrestling with this myself in Linux and found that there currently also isn't a very reliable way to match say a Pulse Audio sink in to a chosen window. (There isn't enough overlapping data to match perfectly it seems, imperfect approximation using heuristics is possible though.) Not to mention the pains of dealing with an unsynced audio stream with a separate video stream. It would require dealing with 2 independent clocks. Playing them side by side would certainly result in desync. Or at the very least require more juggling to get it properly synced up.
> 
> > Hi [@GeorgesStavracas](https://github.com/GeorgesStavracas) ! I would like to work on this issue. Do you still think exposing individual streams is the better approach, or should we aim to provide a single mixed stream?
> 
> So my vote definitely goes to pre-mixed streams. (especially given the context that games have multiple sometimes, you wouldn't want to capture only a single audio stream from a game and miss the rest for example. And you wouldn't _in most cases_ want to put that burden on the user I would say.)
> 
> MacOS has this behavior standardized already, where selecting a single window with the native picker, also includes that window's audio. I think that would be a good example from how it's supposed to work from a user perspective.
> 
> You select a window, and whatever it is that it gets sent to, can choose to use the bundled audio stream or not.

Thanks for the suggestion! I am working on it for the past few days.

**Comment by polyv0x** [comment]
> Thanks for the suggestion! I am working on it for the past few days.

Awesome! As I am only dipping my toes in C development I won't be of much help but I am happy to provide feedback once you have something!
